pallets/phala/src/basepool.rs

```diff
@@ -0,0 +1,645 @@
+			// The nft instance must be wrote to Nft storage at the end of the function
+			// this nft's property shouldn't be accessed or wrote again from storage before set_nft_attr
+			// is called. Or the property of the nft will be overwrote incorrectly.
+			let mut nft = Self::get_nft_attr(base_pool, nft_id)?;
+			// NFT should always settled befroe adding/ removing.
+
+			if let Some(mut settle_method) = maybe_settle_method {
+				settle_method(base_pool, &mut nft, userid.clone());
+			}
+
+			//maybe_settle_nft_slash is only for stakepool
+			//Self.maybe_settle_nft_slash(&mut nft, userid.clone());
+
+			let shares = Self::add_stake_to_new_nft(base_pool, userid.clone(), amount);
+			// Lock the funds
+
+			//ledger should be processed outside basepool (only for stakepool)
+			//Self::ledger_accrue(&userid, a);
+			Self::set_nft_attr(base_pool, nft_id, &nft)
+				.expect("set nft attr should always success; qed.");
+
+			Ok(shares)
+		}
+
+		/// Tries to withdraw a specific amount from a pool.
+		///
+		/// The withdraw request would be delayed if the free stake is not enough, otherwise
+		/// withdraw from the free stake immediately.
+		///
+		/// The updates are made in `pool_info` and `user_info`. It's up to the caller to persist
+		/// the data.
+		///
+		/// Requires:
+		/// 1. The user's pending slash is already settled.
+		/// 2. The pool must has shares and stake (or it can cause division by zero error)
+		#[frame_support::transactional]
+		pub fn push_withdraw_in_queue(
+			base_pool: &mut BasePool<T::AccountId, BalanceOf<T>>,
+			nft: &mut NftAttr<BalanceOf<T>>,
+			nft_id: NftId,
+			userid: T::AccountId,
+			shares: BalanceOf<T>,
+			maybe_vault_pid: Option<u64>,
+		) -> DispatchResult {
+			// Remove the existing withdraw request in the queue if there is any.
+			let (in_queue_nfts, new_withdraw_queue): (VecDeque<_>, VecDeque<_>) = base_pool
+				.withdraw_queue
+				.clone()
+				.into_iter()
+				.partition(|withdraw| withdraw.user == userid);
+			// only one nft withdraw request should be in the queue
+			base_pool.withdraw_queue = new_withdraw_queue;
+			let price = base_pool
+				.share_price()
+				.expect("In withdraw case, price should always exists;");
+
+			match maybe_vault_pid {
+				Some(pid) => {
+					let mut vault = ensure_vault::<T>(pid)?;
+					for withdrawinfo in &in_queue_nfts {
+						let in_queue_nft = Self::get_nft_attr(base_pool, withdrawinfo.nft_id)
+							.expect("get nft attr should always success; qed.");
+						let withdraw_amount = bmul(in_queue_nft.shares.clone(), &price);
+						nft.stakes += in_queue_nft.stakes.min(withdraw_amount);
+						nft.shares += in_queue_nft.shares;
+						Self::burn_nft(base_pool, withdrawinfo.nft_id)
+							.expect("burn nft attr should always success; qed.");
+						vault.basepool.total_stake =
+							vault.basepool.total_stake - in_queue_nft.stakes + withdraw_amount;
+						vault.basepool.free_stake += withdraw_amount;
```

Comment:
Just found out the difference for vault is that you are going to update the `total_value` and `free_stake` of the vault. We can treat the both cases (staker is a vault or user) the same, but additionally take care of the vault later. We can even return these useful information as the function result, and let the upstream (Vault) to process the differential logic.

Reply:
Btw, I'm wondering if the withdrawing stake should be counted as `total_value` or not